Initial commit

tlatmsrud opened the pull request:
테스트가 계속 실패합니다 ㅠㅠ
피드백이 절실합니다.
저를 개조해주세요

tlatmsrud committed:
Initial commit

hannut91 commented on app/src/main/java/com/codesoom/assignment/models/ReqInfo.java:
```diff
@@ -0,0 +1,72 @@
+import java.net.URI;
+import java.util.stream.Collectors;
+
+public class ReqInfo {
```
ReqInfo가 RequestInfo를 줄여서 사용하신거죠. 아주 명확한게 아니라면 줄임말을 사용하지 않는게 좋습니다. 오히려 줄임말을 사용할 때가 좋은 케이스도 있긴 한데, 대표적으로 i, j, k 처럼 특정언어 (C언어)에서는 오히려 더 자세하게 쓰면 안되는 경우가 있습니다.

Golang도 RequestInfo 인스턴스를 r 이렇게 한 글자로 사용하는 관습이 있긴한데, 이건 언어를 사용하는 사람들의 관습을 따라가는게 좋습니다

tlatmsrud replied:
RequestInfo 로 수정하였습니다! 감사합니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/models/ReqInfo.java:
```diff
@@ -0,0 +1,72 @@
+     * @return n번째 path 값
+     * @Desc n번째 path 값을 추출한다.
+     */
+    public String extractPathValue(int n){
+        try{
+            return path[n];
+        }catch(Exception e){
+            return null;
+        }
+
+    }
```
함수가 null을 반환하고 있네요. 이 함수를 사용하는 입장에서는 이 함수가 null을 반환할 수 있다는 것을 알아차리기 어렵고, 또 이 기능을 제공하는 입장에서 "반환값이 null일 수 있으니 null처리를 꼭 해야합니다" 라고 강제할 수 없습니다.

tlatmsrud replied:
pathSegment 배열에서 n번째 값을 가져올 때 OutOfSize 가 발생하는 케이스를 우회하기 위해 넣었던 코드였습니다.
현재는 extractId(), extractFirstPathSegment()  메서드를 만들어 조금 명시적으로 만들어보았는데 뭔가 편치않습니다. ㅠ

hannut91 commented on app/src/main/java/com/codesoom/assignment/models/ReqInfo.java:
```diff
@@ -0,0 +1,72 @@
+     */
+    private void resetPath(HttpExchange exchange){
+        String requestPath = exchange.getRequestURI().getPath();
+        this.path = requestPath.split("/");
```
주어진 path를 `/`를 기준으로 나눌려고 하셨군요. 여러개니까 paths가 되어야 할 것 같습니다. 그리고 URL Path에서 `/`로 나누어진 것 path segment라고 부릅니다. 변수를 사용할 떄 이 변수가 사용되는 도메인 영역에서 사용되는 단어를 사용하시는게 좋습니다

> A path component, consisting of a sequence of path segments separated by a slash (/)

See also
  - https://en.wikipedia.org/wiki/URL

tlatmsrud replied:
path 대신 pathSegments로 수정하였습니다! 감사합니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,153 @@
+
+        // tasks의 value 값들에 대한 Task Collection 추출
+        Collection<Task> tasksValues = tasks.values();
+        OutputStream outputStream = new ByteArrayOutputStream();
+        objectMapper.writeValue(outputStream, tasksValues);
+
+        return outputStream.toString();
```
왜 그냥 값을 쓰지 않고 stream에 쓰는걸까요?

tlatmsrud replied:
생각없이 강의 코드를 사용하다보니 들어갔습니다. stream을 쓰지 않아도 됐었으며, writeValueAsString 메서드를 사용하여 Object를 문자열로 변환하였습니다.

hannut91 replied:
앗 그것보다는 OutputStream을 사용하는 이유는, 한 번에 처리할 수 없는 양이 들어올 경우 컴퓨터가 한 번에 처리할 수 없기 때문에 나눠서 처리하기 위함을 말씀드릴려고 했습니다 ㅎㅎ

tlatmsrud replied:
넵넵 ㅎㅎ 감사합니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,153 @@
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+
+        ReqInfo reqInfo = new ReqInfo(exchange);       // HttpExchange에 대한 요청 객체 생성
+        String method = reqInfo.getMethod();           // 메서드 추출
+        String path = reqInfo.extractPathValue(1);      // 첫번째 path 추출
+        String path2 = reqInfo.extractPathValue(2);     // 두번째 path 추출
+        String reqBody = reqInfo.getBody();            // 요청 바디 추출
+        String resBody = "";
```
RegInfo가 무언가를 하는 객체라기 보다 값을 처리하는 처리기처럼 보이네요. 객체 스스로가 무언가 할 수 있도록 역할이 있어야 할 것 같아요

tlatmsrud replied:
RequestInfo 객체가 요청 값에 대한 책임을 느낄(?)정도의 기능을 추가해보았습니다.
 1) validation Check
 2) id 값 존재 여부 체크. 기존 로직에서는 핸들러에서 pathSegments[2] != null 로 체크하였음.
 3) id 값 추출 - Long으로 캐스팅 후 리턴
역할 이라는 것에 대한 예시나 힌트를 주실 수 있으실까요? 확 와닿고 싶습니다.

tlatmsrud committed:
1차 피드백

1. RequestInfo 객체가 요청 값에 대한 책임을 느낄(?)정도의 기능을 추가해보았습니다.
 1) validation Check
 2) id 값 존재 여부 체크. 기존 로직에서는 핸들러에서 pathSegments[2] != null 로 체크하였음.
 3) id 값 추출 - Long으로 캐스팅 후 리턴
2. stream을 쓰지 않아도 됐었으며, writeValueAsString 메서드를 사용하여 Object를 문자열로 변환하였습니다.
3. 넵 줄여서 사용했던겁니다. RequestInfo로 수정하였습니다.
4. path 대신 pathSegments로 수정하였습니다
5. pathSegment 배열에서 n번째 값을 가져올 때 OutOfSize 가 발생하는 케이스를 우회하기 위해 넣었던 코드였습니다.
현재는 extractId(), extractFirstPathSegment()  메서드를 만들어 조금 명시적으로 만들어보았는데 뭔가 편치않습니다. ㅠ

궁금한 사항
1. 현재 GET, POST, tasks, HTTP 상태코드와 같은 값들이 모두 하드코딩되어있습니다. 이러한 값들은 enum으로 처리하는게 더 나을까요?
2. 1번 피드백 내용 중 '역할'이라는 것에 대한 예시를 들어주실 수 있으신가요? 확 와닿고싶습니다
3. RequestInfo 클래스에 pathSegment 의 n번째 값을 가져오는 것 대신 extractId, extractFirstPathSegment를 쓰고 있는게 걸립니다.
요청 pathSegment 의 값이 하나만 추가되어도 코드를 갈아엎어야하는 상황이라.. 혹시 더 좋은 방법이 있을까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,26 @@
     public static void main(String[] args) {
+
+        try{
+            InetSocketAddress address = new InetSocketAddress(8000);
```
8000은 `포트번호`를 의미하고 있는 것 같아요. 8000이 단순히 숫자가 아니라 특별한 의미를 갖는다는 것을 상수로 표현해보는 것은 어떨까요?

See also

- [Replace Magic Literal](https://refactoring.com/catalog/replaceMagicLiteral.html)

tlatmsrud replied:
port 값을 상수처리하였습니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,140 @@
+    public void handle(HttpExchange exchange) throws IOException {
+
+        try{
+            RequestInfo requestInfo = new RequestInfo(exchange);        // HttpExchange에 대한 요청 객체 생성
+            String method = requestInfo.getMethod();                    // 메서드 추출
+            String reqBody = requestInfo.getBody();                     // 요청 바디 추출
+            String path = requestInfo.extractFirstPathSegment();        // 첫번째 세그먼트 추출
+
+            // task 조회 - GET /tasks 또는 GET/tasks/{id}
+            if("GET".equals(method) && "tasks".equals(path)){
```
RequestInfo가 exchange로부터 method, path를 만들어내는 이유는 이 요청이 Task조회 인지, Task 생성인지 판단하기 위해서 만들어내고 있는 것 같아요. 어떤 요청인지 파악하는 일을 누군가에게 맡기고 싶은데, 누가 잘할지 생각해보면, method와 path를 알고있는 RequestInfo가 잘할 것 같죠?

tlatmsrud replied:
요청 값에 대한 vaildation 체크 기능과 요청을 분석하는 기능을 추가하였습니다.
path segment의 'tasks' 를 체크하는 로직이 겹쳐 해당 로직은 유효성 체크 메서드에 넣었으며,
요청을 분석하는 메서드에서 어떤 명령을 처리해야하는지 enum 형태로 관리해보았습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,140 @@
+            if("POST".equals(method) && "tasks".equals(path)){
+
+                Task task = toTask(reqBody);
+                task.setId(++taskId);
```
전위연산자를 사용해서 값이 변경되기 전의 값을 할당하고 나서 값을 증가시키고 있는 것 같아요. 사용법을 정확히 알고 있다면 ++value 와 value++ 을 구분해서 사용할 수 있겠지만, 사실 조금 헷갈릴수도 있는 문법이에요. 우리가 한 번 더 생각해야되는거죠. 아주 조금이라도 헷갈릴 수 있거나 다르게 해석될 여지가 있다면 저는 사용을 지양하는 것이 좋다고 생각해요.

tlatmsrud replied:
전위 연산자 대신 더하기 연산자를 사용하였으며 increaseTaskId 메서드로 만들었습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/models/RequestInfo.java:
```diff
@@ -0,0 +1,105 @@
+        return false;
+    }
+
+    /**
+     * @Desc path segments 중 id에 해당하는 값을 추출한다.
+     */
+    public Long extractId(){
+
+        return Long.parseLong(pathSegments[2]);
+
+    }
+
+    /**
+     * @Desc path segments 중 id에 해당하는 값을 추출한다.
+     * @Feedback
+     *  1. 함수를 사용하는 입장에서는 이 함수가 null을 반환하거나, null 체크를 해야한다는 것을 알아차릴 수 없음.
+     */
+    public String extractFirstPathSegment(){
+
+        return pathSegments[1];
+
+    }
```
여기서 왜 1이고 왜 2인지 알기 굉장히 어렵습니다. enum으로 추출한다고 해서 더 나아질 것 같지는 않아요. 다른 대안이 있을까요?

tlatmsrud replied:
extractFirstPathSegment 메서드가 필요했던 이유는 핸들러 클래스에서 path Segment 값을 비교하기 위함이었습니다.
이 과정을 RequestInfo에게 위임했기 때문에 해당 메서드는 삭제하였습니다.
extractId 또한 핸들러 클래스에서 path Segment 값을 추출하기 위함이었습니다.
이 과정은 RequestInfo 생성자 메서드에서 선 수행 후 멤버필드에 넣어줬습니다.
id 값이 null일 수 있다는 것을 암시하기 위해 멤버필드 초기화 단계에 id = null 구문을 넣어봤습니다.

tlatmsrud committed:
두번째 피드백 반영

1. 요청에 대한 유효성체크 및 분석을 RequestInfo에게 위임
2. 전위연산자 대신 더하기 연산을 사용
3. firstPathSegment 값을 handler에서 호출하지 않도록 하여 개발자로
   하여금 pathSegment[1] 을 의식하지 않도록 하였음.
4. 생성자 메서드 내에서 id를 초기화하였으며, handler에서는 getter를 통해
   가져오도록 함.
5. uri의 tasks에 대한 유효성 체크를 생성자 메서드에서 수행함으로써, 요청
   값을 분석하는 로직에 있던 중복된 tasks 문자열 체크를 제거하였음.

hannut91 commented on app/src/main/java/com/codesoom/assignment/enums/Command.java:
```diff
@@ -0,0 +1,10 @@
+package com.codesoom.assignment.enums;
+
+public enum Command {
+
```
여기 빈 줄이 하나 들어가 있네요

tlatmsrud replied:
빈줄 제거하였습니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,144 @@
+            Long id = requestInfo.getId();
+
+            // task 리스트 조회
+            if(command == Command.GET_TASK_LIST) {
```
요청 정보에서 커맨드를 얻은 후에 여기서 비교해주고 있네요. 객체에게 무언가를 요청하기 보다는 시켜야 합니다. 무언가 물어보고 내가 무언가를 한다는 것은 객체가 굉장히 수동적이게 됩니다. 우리는 묻지말고 시키기만 해야합니다.
```java
if (requestInfo.isGetRequet()) {
   // ... 생략
}
```

tlatmsrud replied:
Command 값을 체크하는 로직을 RequestInfo 객체 내에 넣었습니다.  생각해보니 지금은 handler에서만 사용하지만 다른 객체에서도 사용해야 하는 상황이 발생하면 동일한 체크 로직이 들어가게 되어 문제가 발생할 것 같습니다.
Command에 대한 값을 받아오는게 아니라 체크 까지 모두 한 결과를 받아오는 메서드를 만들었습니다.!
감사합니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,144 @@
+     * TaskId 증가
+     */
+    private void increaseTaskId(){
+        TASKS_ID = TASKS_ID + 1;
```
자바에는 UPPER_SNAKECASE는 static final에만 사용하는 관습이 있습니다. 일반적인 자바개발자가 이것을 보면 혼란스러울 수 있을 것 같아요

tlatmsrud replied:
넵넵 카멜케이스로 변경하였습니다. 감사합니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/models/RequestInfo.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private void requestAnalysis(HttpExchange exchange) {
+        String      method = exchange.getRequestMethod();
+        String      path = exchange.getRequestURI().getPath();
+        String[]    pathSegments = path.split("/");
+
+        this.id      = extractId(pathSegments);
+        this.command = resetCommand(method);
```
일반적인 자바개발자는 이렇게 정리하지 않는 것 같아요. IDEA의 자동 정렬 기능을 사용해 보세요. 혹은 파일을 저장할 떄 마다 자동으로 정렬되도록 하는 기능도 있습니다.

참고로 Terraform이나 Golang은 위와 같이 정렬하는 것을 선호합니다. 언어를 사용하는 생태계에 따라 달라지는 것이니 주의하시면 좋을 것 같아요. 다른 곳에서 사용하는 관습을 그대로 적용하기 보다는 어떻게 사용되고 있는지 살펴보는 것도 좋겠습니다

tlatmsrud replied:
넵 감사합니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/models/RequestInfo.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+
+    public Long extractId(String[] pathSegments){
+
+        if(pathSegments.length == 3){
+            return Long.parseLong(pathSegments[LOCATION_OF_ID]);
+        }
+        return null;
+    }
+
```
pathSegment의 자리로 id를 추출하고 있군요. 정규식을 사용해서 id를 얻어 보세요.

See also
  - https://www.bsidesoft.com/8232

tlatmsrud replied:
정규식을 사용해서 id를 얻는 방법으로 수정하였습니다. 이렇게 하니 pathSegments를 추출할 필요가 없어 split 하는 부분도 제거하였습니다.

tlatmsrud committed:
객체에게 묻지말고 시켜라

멤버필드의 값을 그냥 달라고 묻지 않고, 객체 내 정보가 이게 맞는지
체크해주세요 라고 시켜보았습니다. 객체야 니 일은 니가 알아서 해야지 내가
대신해준다면 나중에 일할때 넌 나한테 의존적이게될거야. 독립적으로
살란말야

tlatmsrud committed:
private 를 public으로 변경하라

tlatmsrud committed:
Request에 대한 체크는 requestInfo에게 맡겨라

tlatmsrud committed:
UPPER_SNAKECASE를CAMELCASE로 수정하라

자바에서 UPPER_SNAKECASE 는 static final에만 사용하는 관습이 있기
때문이다.

tlatmsrud committed:
IDE의정렬기능을 사용하여 코드를 정렬하라

개발 언어마다 코드를 정렬하는 방법이 다르다. 내가 특정부분에 사용했던
정렬 방식은 Golang이나 Terraform에서 선호하는 방식이라고 한다. 일반적인
정렬 방법을 사용하기 위해 IDE의 정렬기능을 사용하였다.

tlatmsrud committed:
ID 추출을 정규식으로 구현하라

pathSegment 의 id 추출 방법을 정규식으로 수정함. 정규식은
EXTRACT_ID_REGULAR_EXPRESSION 이라는 static final 변수로 설정함.
